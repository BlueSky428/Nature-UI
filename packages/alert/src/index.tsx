import * as React from 'react';
import clx from 'clsx';
import { FiInfo, FiAlertCircle, FiAlertTriangle, FiCheckCircle } from 'react-icons/fi';

const SUBTLE_TEXT = 'text-gray-800';

export const ALERT_STATUSES = {
  error: {
    bg: 'bg-red-200',
    iconColor: 'text-red-600 mr-3',
    icon: FiAlertCircle,
    variant: {
      solid: 'bg-red-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  info: {
    bg: 'bg-blue-200',
    iconColor: 'text-blue-600 mr-3',
    icon: FiInfo,
    variant: {
      solid: 'bg-blue-600',
      subtle: SUBTLE_TEXT,
    },
  },
  success: {
    bg: 'bg-green-200',
    iconColor: 'text-green-600 mr-3',
    icon: FiCheckCircle,
    variant: {
      solid: 'bg-green-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  warning: {
    bg: 'bg-orange-200',
    iconColor: 'text-orange-600 mr-3',
    icon: FiAlertTriangle,
    variant: {
      solid: 'bg-orange-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
};

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof ALERT_STATUSES;
  /**
   * The variant of the alert style to use
   */
  variant?: 'solid' | 'subtle';

  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;

  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ElementType;
}

export type AlertProps = AlertOptions;

const BASE_STYLE = 'px-4 py-3 flex items-center';

const Alert: React.ForwardRefExoticComponent<AlertProps> = React.forwardRef(
  (props: AlertProps, ref: React.Ref<any>) => {
    const {
      className = '',
      status = 'success',
      children,
      variant = 'subtle',
      component: Component = 'div',
    } = props;

    // const Component = 'div';

    const VARIANT: string = ALERT_STATUSES[status].variant[variant];

    const componentClass = clx(BASE_STYLE, {
      [className]: className,
      [ALERT_STATUSES[status].bg]: status,
      [VARIANT]: variant,
    });

    const Icon = ALERT_STATUSES[status];
    const iconClasses = clx({
      [VARIANT]: variant,
      'mr-3': variant,
      [Icon.iconColor]: variant !== 'solid',
    });

    const IconComponent = Icon.icon;

    return (
      <Component className={componentClass} ref={ref}>
        {status && <IconComponent className={iconClasses} size={20} />}
        <span>{children ? children : 'This is an alert ðŸ™‚'}</span>
      </Component>
    );
  }
);

Alert.displayName = 'Alert';

export default Alert;
