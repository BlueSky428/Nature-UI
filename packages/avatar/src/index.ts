export * from './avatar';
export * from './avatarGroup';
