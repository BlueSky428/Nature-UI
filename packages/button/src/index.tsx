import * as React from 'react';
import clx from 'clsx';

interface ButtonProps {
  /**
   * The label of the button
   */
  'aria-label'?: string;
  /**
   * The text color of the button. Use a color key passed in theme.colors.
   */
  text?: string;
  /**
   * The background color of the button. Use a color key passed in theme.colors.
   */
  color?: string;
  /**
   * The variant of the button style to use.
   */
  variant?: 'outline' | 'ghost' | 'link' | 'solid' | 'none';
  /**
   * If true, the button will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If true, the button will show a spinner.
   */
  isLoading?: boolean;
  /**
   * The label to show in the button when isLoading is true. If no text is passed, it only shows the spinner
   */
  loadingText?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'none';
  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;
  /**
   * The lightness or how dark you want the background color to be
   */
  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   * defaults to "button"
   */
  component?: React.ElementType;
  id?: string;
  /**
   *  when the href prop is set it changes the tag name to a
   */
  href?: string;
  onClick?: () => any;
}

const darken = (value: string): string => {
  const splitStr = value.split('-');
  const dept = Number(splitStr[1]);
  const color = splitStr[0];

  if (dept >= 800) {
    return `${color}-${dept - 200}`;
  }

  return `${color}-${dept + 200}`;
};

const lighten = (value: string): string => {
  const color = value.split('-')[0];

  return `${color}-${100}`;
};

const Button = React.forwardRef((props: ButtonProps, ref: React.Ref<any>) => {
  const {
    component: Component = 'button',
    variant = 'solid',
    color = 'teal-500',
    text = 'white',
    size = 'md',
    children,
    ...rest
  } = props;
  const DEFAULT_CLASS = 'block focus:shadow-outline focus:outline-none rounded font-semibold flex';
  const STYLES = {
    solid: `bg-${color} text-${text} hover:bg-${darken(color)}`,
    outline: `bg-transparent hover:bg-${lighten(
      text
    )} text-${text} border border-${text} focus:border-transparent`,
    ghost: `hover:bg-${lighten(text)} text-${text}`,
    link: `hover:underline text-${text}`,
    md: 'px-4 py-2',
  };

  let BTNClass: string;

  if (variant === 'none') {
    BTNClass = '';
  } else {
    BTNClass = clx(DEFAULT_CLASS, {
      [STYLES[size]]: size && variant !== 'link',
      [STYLES[variant]]: variant,
    });
  }

  return (
    <Component className={BTNClass} ref={ref} {...rest}>
      {children ? children : 'Button'}
    </Component>
  );
});

export default Button;
