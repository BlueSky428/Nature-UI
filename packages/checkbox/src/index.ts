export * from './checkbox';
export * from './checkbox-group';
export * from './checkbox-types';
export * from './use-checkbox';
export * from './use-checkbox-group';
