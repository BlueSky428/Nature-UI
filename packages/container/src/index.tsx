import * as React from 'react';
import clx from 'clsx';

interface IContainer {
  children?: React.ReactNode;
  /**
   * Determine the max-width of the container.
   */
  size?: 'lg' | 'md' | 'sm' | 'xl' | 'xs';
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ComponentType;
  className?: string;
}

const BASE_STYLE = 'mx-auto container';
const LG = 'px-20';
const MD = 'px-32';
const SM = 'px-48';
const XL = 'px-0';
const XS = 'px-64';

const Container: React.ForwardRefExoticComponent<IContainer> = React.forwardRef(
  (props: IContainer, ref: React.Ref<any>) => {
    const { component: Component = 'div', size = 'lg', children, className = '' } = props;

    const componentClass: string = clx(BASE_STYLE, {
      [LG]: size === 'lg',
      [MD]: size === 'md',
      [SM]: size === 'sm',
      [XL]: size === 'xl',
      [XS]: size === 'xs'
    });

    return (
      <Component className={componentClass} ref={ref}>
        <Component className={className}>{children}</Component>
      </Component>
    );
  }
);

Container.displayName = 'Container';

export default Container;
