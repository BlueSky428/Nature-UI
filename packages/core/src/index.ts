export * from '@nature-ui/alert';
export * from '@nature-ui/alert-dialog';
export * from '@nature-ui/avatar';
export * from '@nature-ui/breadcrumb';
export * from '@nature-ui/button';
export * from '@nature-ui/checkbox';
export * from '@nature-ui/close-button';
export * from '@nature-ui/container';
export * from '@nature-ui/counter';
export * from '@nature-ui/drawer';
export * from '@nature-ui/form-control';
export * from '@nature-ui/hooks';
export * from '@nature-ui/icon';
export * from '@nature-ui/image';
export * from '@nature-ui/input';
export * from '@nature-ui/layout';
export * from '@nature-ui/modal';
export * from '@nature-ui/popover';
export * from '@nature-ui/popper';
export * from '@nature-ui/portal';
export * from '@nature-ui/progress';
export * from '@nature-ui/radio';
export * from '@nature-ui/spinner';
export * from '@nature-ui/system';
export * from '@nature-ui/toast';
export * from '@nature-ui/transition';
export * from '@nature-ui/visually-hidden';
