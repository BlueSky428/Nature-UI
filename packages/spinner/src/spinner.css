.spinner {
  border-radius: 100%;

  animation: spin 0.5s infinite linear;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
