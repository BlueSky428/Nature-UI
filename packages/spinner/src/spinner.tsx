import * as React from 'react';
import { nature, PropsOf, clsx, css, keyframes } from '@nature-ui/system';
import VisuallyHidden from '@nature-ui/visually-hidden';
import { __DEV__ } from '@nature-ui/utils';

interface SpinnerOptions {
  /**
   * The color of the spinner
   */
  color?: string;
  /**
   * The thickness of the spinner
   * @example
   * ```jsx
   * <Spinner thickness="4px"/>
   * ```
   */
  thickness?: string;
  size?: number | 'xs' | 'sm' | 'md' | 'lg';
  label?: string;
}

const SIZES = {
  xs: 4,
  sm: 6,
  md: 8,
  lg: 12,
};

const SpinnerComp = nature('span');

export type SpinnerType = PropsOf<typeof SpinnerComp>;

export const Spinner = React.forwardRef(
  (props: SpinnerType & SpinnerOptions, ref: React.Ref<HTMLDivElement>) => {
    const {
      className = '',
      thickness = '2px',
      color = 'blue-500',
      size = 'xs',
      label = 'Loading...',
      ...rest
    } = props;

    const _css = css`
      border-radius: 100%;

      animation: spin 0.5s infinite linear;
    `;

    const spin = keyframes`
      from {
        transform: rotate(0deg);
      }
      to {
        transform: rotate(360deg);
      }
    `;

    const _size = typeof size === 'string' ? SIZES[size] : size;
    const DEFAULTS = `inline-block overflow-hidden border-2 border-transparent border-t-2 align-middle w-${_size} h-${_size}`;

    const _classNames = clsx(className, _css, spin, DEFAULTS, {
      [`text-${color}`]: color,
    });

    return (
      <SpinnerComp
        className={_classNames}
        style={{
          borderTop: `${thickness} solid currentColor`,
          borderLeft: `${thickness} solid currentColor`,
        }}
        ref={ref}
        {...rest}
      >
        {label && <VisuallyHidden>{label}</VisuallyHidden>}
      </SpinnerComp>
    );
  },
);

if (__DEV__) {
  Spinner.displayName = 'Spinner';
}
