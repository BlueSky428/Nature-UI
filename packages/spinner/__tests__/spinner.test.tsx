import { render } from '@nature-ui/test-utils';
import { Spinner } from '../src';

test('Spinner renders correctly', () => {
  const { asFragment } = render(<Spinner />);

  expect(asFragment()).toMatchSnapshot();
});
