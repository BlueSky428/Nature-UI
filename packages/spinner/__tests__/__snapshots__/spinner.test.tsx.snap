// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner renders correctly 1`] = `
<DocumentFragment>
  <span
    class="animation-1nvx475 border-2 w-4 h-4 inline-block overflow-hidden border-transparent border-t-2 align-middle text-blue-500 css-hxvqwd"
    style="border-top: 2px solid currentColor; border-left: 2px solid currentColor;"
  >
    <span
      class="css-e941bk"
    >
      Loading...
    </span>
  </span>
</DocumentFragment>
`;
